src/Cassandra/Cluster.cs

```diff
@@ -242,13 +245,50 @@ public async Task<ISession> ConnectAsync(string keyspace)
         async Task<TSession> IInternalCluster.ConnectAsync<TSession>(ISessionFactory<TSession> sessionFactory, string keyspace)
         {
             await Init().ConfigureAwait(false);
-            var session = await sessionFactory.CreateSessionAsync(keyspace, _serializer).ConfigureAwait(false);
+            var newSessionName = await GetNewSessionNameAsync().ConfigureAwait(false);
+            var session = await sessionFactory.CreateSessionAsync(keyspace, _serializer, newSessionName).ConfigureAwait(false);
             await session.Init().ConfigureAwait(false);
             _connectedSessions.Add(session);
             Cluster.Logger.Info("Session connected ({0})", session.GetHashCode());
             return session;
         }
 
+        private async Task<string> GetNewSessionNameAsync()
+        {
+            var sessionCounter = await GetAndIncrementSessionCounterAsync().ConfigureAwait(false);
+            if (sessionCounter == 0 && Configuration.SessionName != null)
+            {
+                return Configuration.SessionName;
+            }
+
+            var prefix = Configuration.SessionName ?? Configuration.DefaultSessionName;
+            return prefix + sessionCounter;
+        }
+
+        private async Task<long> GetAndIncrementSessionCounterAsync()
+        {
+            await _sessionCounterLock.WaitAsync().ConfigureAwait(false);
```

Comment:
This might be a bit nitpicky but I'm concerned that someone out there might be constantly creating sessions from the same cluster instance so I'd rather not impose a set max number of sessions. I'm +1 on removing locks though. I've changed this, lmk your thoughts.

Reply:
lgtm!